src/Conversion/ONNXToKrnl/Math/Elementwise.cpp

```diff
@@ -1452,21 +1571,36 @@ struct ONNXElementwiseUnaryOpLowering
     if (enableSIMD && !isScalar && !hasNonIdentityLayout(operands)) {
       int64_t simdUnroll =
           canBeVectorized<ONNXUnaryOpShapeHelper, ElementwiseUnaryOp>(
-              shapeHelper, create, memRefType);
+              shapeHelper, create, outputMemRefType);
       if (simdUnroll > 0)
         return getUnaryBinarySimdCodeFullyFlattened<ElementwiseUnaryOp>(
-            rewriter, create, &shapeHelper, op, memRefType, operands, alignment,
-            simdUnroll);
+            rewriter, create, &shapeHelper, op, outputMemRefType, operands,
+            alignment, simdUnroll);
+    }
+
+    // Try to fuse the unary elementwise consumers
+    SmallVector<Operation *, 2> fusibleOps;
+    SmallVector<EmitScalarFunc, 2> fuseEmitFunctions;
+    bool fusibleOpsFound =
+        findFusibleUnaryOps(op, fusibleOps, fuseEmitFunctions);
+    if (fusibleOpsFound) {
+      // After fusion, the only store is for the last Op.
+      // Therefore, the allocation should be the output of the last Op
+      Operation *lastOp = fusibleOps[fusibleOps.size() - 1];
+      outputMemRefType =
+          this->typeConverter->convertType(lastOp->getResults()[0].getType())
+              .template cast<MemRefType>();
```

Comment:
We use these code repeatedly. I think it can be simpler if we have a class, say, `UnaryOpFusionHelper` and hide all `fusibleOps`, `fuseEmitFunctions`, and `outputMemRefType` inside, e.g.
```C
UnaryOpFusionHelper fusionHelper(rewriter, loc);
fusionHelper.fuse(op);
bool fusibleOpsFound = fusionHelper.isFeasible();
outputMemRefType = fusionHelper.getOutputType();
```

and inside `krnl.iterate`:
```C
loweredOpResult = fusionHelper.emitFuseOps(loweredOpResult);
```

Reply:
It would be great to do so, but I am afraid that it requires a lot of changes at this time, since it depends on `emitScalarForOp` that should be reorganized too.

Reply:
You are right. The code of template for emitScalarForOp has to be reorganized. I pushed the change that define the OpFusionHelper inside the Elementwise.cpp.